### DISCUSSION TITLE
Using file signatures to identify executables, instead of extensions.

### DISCUSSION THREAD

**Discussion by daschewie** [pweingar/FoenixMCP#15] [Ideas]
# Proposal

PGX, PGZ, and ELF binaries have a file signature in the header to identify them as executables. I think it would be nice to support executables without file extensions, because it would hide the implementation details, and save us from typing extensions.

What does the community think?

**Comment by pweingar** [comment]
I was thinking about something like this, where we wouldn't be tied to extensions... my concern though was that I don't think we have an attribute bit to serve as an executable flag. Maybe we could use the archive attribute bit, since no one really uses that anyway. Or maybe it's not an issue after all. The other concern is that PGZ really only has "z" and "Z" as the signature, which maybe isn't the most specific. But maybe that also isn't really a big deal.

**Comment by vinz6751** [comment]
I support the idea, but as PJG points out, having a single letter can be risky.
I think the loader could be a structure with function pointers e.g. to get the list of supported extensions (.pgz, .pgx, .elf, "none") and "`bool can_run(channel)`". Please note that there can be many, e.g. for the Atari TOS loader there can be .PRG .TOS, .TTP, .APP, .GTP !.  

The CLI/shell can also help. E.g. if you just type "myprog" as shortcut for "myprog.pgz", the CLI searches for myprog.* , find myprog.pgz then checks if any loader supports ".pgz", then calls the loader. Of course if you have myprog.pgx myprog.pgz, which program is launched will depend on the order in which loaders are interrogated.

The idea about using the archive bit means that if we want to use filesystem attributes to identify executable files, the file system should support it ; but FAT doesn't out of the box. So rather than bending the archive bit we should add another bit if possible, or otherwise use another filesystem (out of the question I suppose!). At first glance there seem to be 2 "reserved" bit after flags, I'm not sure if they have an unofficial use or they're free for use.

**Reply by daschewie** [reply]
Yeah, I am starting to agree with the notion that file signatures may not be the best idea. The wildcard search has been brought up by Peter in the past. I think the wildcard with a PATH environment variable maybe the best bet.
